File: uikit/src/main/java/adapter/GroupMemberAdapter.java
package adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.cometchat.pro.constants.CometChatConstants;
import com.cometchat.pro.core.CometChat;
import com.cometchat.pro.models.GroupMember;
import com.cometchat.pro.models.TypingIndicator;
import com.cometchat.pro.uikit.R;
import com.cometchat.pro.uikit.databinding.UserListRowBinding;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import utils.FontUtils;
import utils.Utils;

/**
 * Purpose - GroupMemberAdapter is a subclass of RecyclerView Adapter which is used to display
 * the list of group members. It helps to organize the list data in recyclerView.
 * It also help to perform search operation on list of groups members.
 *
 * Created on - 20th December 2019
 *
 * Modified on  - 24th January 2020
 *
 */


public class GroupMemberAdapter extends RecyclerView.Adapter<GroupMemberAdapter.GroupMemberViewHolder> {

    private Context context;

    private String groupOwnerId;

    private static final String TAG = GroupMemberAdapter.class.getSimpleName();

    private List<GroupMember> groupMemberList = new ArrayList<>();

    private FontUtils fontUtils;

    /**
     * It is a constructor which is used to initialize wherever we needed.
     *
     * @param context is a object of Context.
     */
    public GroupMemberAdapter(Context context) {
        this.context = context;
        fontUtils=FontUtils.getInstance(context);
    }

    /**
     * It is constructor which takes groupMemberList as parameter and bind it with groupMemberList in adapter.
     *
     * @param context is a object of Context.
     * @param groupMemberList is a list of group member used in this adapter.
     */
    public GroupMemberAdapter(Context context, List<GroupMember> groupMemberList, String groupOwnerId) {
        this.groupMemberList = groupMemberList;
        this.groupOwnerId = groupOwnerId;
        this.context = context;
        fontUtils=FontUtils.getInstance(context);

    }

    @NonNull
    @Override
    public GroupMemberViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int i) {

        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());

        UserListRowBinding userListRowBinding = DataBindingUtil.inflate(layoutInflater, R.layout.user_list_row, parent, false);

        return new GroupMemberViewHolder(userListRowBinding);
    }

    /**
     *  This method is used to bind the GroupMemberViewHolder contents with groupMember at given
     *  position. It set name icon, scope with respective GroupMemberViewHolder content.
     *
     * @param groupMemberViewHolder is a object of GroupMemberViewHolder.
     * @param i is a position of item in recyclerView.
     * @see GroupMember
     */
    @Override
    public void onBindViewHolder(@NonNull GroupMemberViewHolder groupMemberViewHolder, int i) {

        GroupMember groupMember = groupMemberList.get(i);

        groupMemberViewHolder.userListRowBinding.avUser.setBackgroundColor(context.getResources().getColor(R.color.colorPrimary));
        groupMemberViewHolder.userListRowBinding.getRoot().setTag(R.string.user, groupMember);


        if (groupMember.getUid().equals(CometChat.getLoggedInUser().getUid())) {
            groupMemberViewHolder.userListRowBinding.txtUserName.setText(R.string.you);
        } else
            groupMemberViewHolder.userListRowBinding.txtUserName.setText(groupMember.getName());

        if (groupOwnerId != null && groupMember.getUid().equals(groupOwnerId) &&
                groupMember.getScope().equals(CometChatConstants.SCOPE_ADMIN)) {
            groupMemberViewHolder.userListRowBinding.txtUserScope.setText(R.string.owner);
        } else if (groupMember.getScope().equals(CometChatConstants.SCOPE_ADMIN)) {
            groupMemberViewHolder.userListRowBinding.txtUserScope.setText(R.string.admin);
        } else if (groupMember.getScope().equals(CometChatConstants.SCOPE_MODERATOR)){
            groupMemberViewHolder.userListRowBinding.txtUserScope.setText(R.string.moderator);
        } else {
            groupMemberViewHolder.userListRowBinding.txtUserScope.setText("");
        }

        groupMemberViewHolder.userListRowBinding.txtUserName.setTypeface(fontUtils.getTypeFace(FontUtils.robotoMedium));
        if (groupMember.getAvatar() == null || groupMember.getAvatar().isEmpty())
            groupMemberViewHolder.userListRowBinding.avUser.setInitials(groupMember.getName());
        else
            groupMemberViewHolder.userListRowBinding.avUser.setAvatar(groupMember.getAvatar());

        if (Utils.isDarkMode(context))
        {
            groupMemberViewHolder.userListRowBinding.txtUserName.setTextColor(context.getResources().getColor(R.color.textColorWhite));
            groupMemberViewHolder.userListRowBinding.tvSeprator.setBackgroundColor(context.getResources().getColor(R.color.grey));
            groupMemberViewHolder.userListRowBinding.txtUserScope.setTextColor(context.getResources().getColor(R.color.textColorWhite));
        } else {
            groupMemberViewHolder.userListRowBinding.txtUserName.setTextColor(context.getResources().getColor(R.color.primaryTextColor));
            groupMemberViewHolder.userListRowBinding.tvSeprator.setBackgroundColor(context.getResources().getColor(R.color.light_grey));
            groupMemberViewHolder.userListRowBinding.txtUserScope.setTextColor(context.getResources().getColor(R.color.secondaryTextColor));
        }

        if (i == getItemCount() - 1)
            groupMemberViewHolder.userListRowBinding.tvSeprator.setVisibility(View.GONE);


    }

    @Override
    public int getItemCount() {
        return groupMemberList.size();

    }

    /**
     * This method is used to add group members in groupMemberList of adapter
     *
     * @param groupMembers is a list of group members which will be added in adapter.
     * @see GroupMember
     */
    public void addAll(List<GroupMember> groupMembers) {

        for (GroupMember groupMember : groupMembers) {
            if (!groupMemberList.contains(groupMember)) {
                groupMemberList.add(groupMember);
            }
        }
        notifyDataSetChanged();
    }

    /**
     * This method is used to set search list in a groupMemberList of adapter.
     *
     * @param filterlist is a list of searched group members.
     */
    public void searchGroupMembers(List<GroupMember> filterlist) {
        this.groupMemberList = filterlist;
        notifyDataSetChanged();
    }

    /**
     * This method is used to add group member in a groupMemberList of adapter.
     *
     * @param joinedUser is object of GroupMember which will be added in a groupList.
     * @see GroupMember
     */
    public void addGroupMember(GroupMember joinedUser) {
        groupMemberList.add(joinedUser);
        notifyDataSetChanged();
    }

    /**
     * This method is used to remove group member from groupMemberList of adapter.
     *
     * @param groupMember is a object of GroupMember which will be removed from groupList.
     * @see GroupMember
     */
    public void removeGroupMember(GroupMember groupMember) {
         if (groupMemberList.contains(groupMember)) {
             groupMemberList.remove(groupMember);
             notifyDataSetChanged();
         }
    }

    /**
     * This method is used to update group member from a groupMemberList of a adapter.
     *
     * @param groupMember is a object of GroupMember which will updated with old group member in
     *                    groupMemberList.
     * @see GroupMember
     */
    public void updateMember(GroupMember groupMember) {
        if (groupMemberList.contains(groupMember)) {
            int index = groupMemberList.indexOf(groupMember);
            groupMemberList.remove(groupMember);
            groupMemberList.add(index, groupMember);
            notifyItemChanged(index);
        }
    }

    public void resetAdapter() {
        groupMemberList.clear();
        notifyDataSetChanged();
    }

    class GroupMemberViewHolder extends RecyclerView.ViewHolder {

        UserListRowBinding userListRowBinding;

        GroupMemberViewHolder(UserListRowBinding userListRowBinding) {
            super(userListRowBinding.getRoot());
            this.userListRowBinding = userListRowBinding;

        }
    }

    /**
     * This method is used to update group members in a groupMemberList of a adapter.
     *
     * @param groupMembers is a list of updated group members.
     */
    public void updateGroupMembers(List<GroupMember> groupMembers) {
        for (int i = 0; i < groupMembers.size(); i++) {
            if (groupMemberList.contains(groupMembers.get(i))) {
                int index = groupMemberList.indexOf(groupMembers.get(i));
                groupMemberList.remove(index);
                groupMemberList.add(index, groupMembers.get(i));
            } else {
                groupMemberList.add(groupMembers.get(i));
            }
        }
        notifyDataSetChanged();
    }
}

File: uikit-kotlin/src/main/java/com/cometchat/pro/uikit/ui_components/groups/group_members/GroupMemberAdapter.kt
package com.cometchat.pro.uikit.ui_components.groups.group_members

import com.cometchat.pro.uikit.ui_components.groups.group_members.GroupMemberAdapter.GroupMemberViewHolder
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.cometchat.pro.constants.CometChatConstants
import com.cometchat.pro.core.CometChat
import com.cometchat.pro.models.GroupMember
import com.cometchat.pro.uikit.R
import com.cometchat.pro.uikit.databinding.CometchatUserListItemBinding
import com.cometchat.pro.uikit.ui_resources.utils.FontUtils
import com.cometchat.pro.uikit.ui_resources.utils.Utils
import java.util.*

/**
 * Purpose - GroupMemberAdapter is a subclass of RecyclerView Adapter which is used to display
 * the list of group members. It helps to organize the list data in recyclerView.
 * It also help to perform search operation on list of groups members.
 *
 * Created on - 20th December 2019
 *
 * Modified on  - 24th January 2020
 *
 */
class GroupMemberAdapter : RecyclerView.Adapter<GroupMemberViewHolder> {
    private lateinit var context: Context
    private var groupOwnerId: String? = null
    private var groupMemberList: MutableList<GroupMember> = ArrayList()
    private lateinit var fontUtils: FontUtils


    /**
     * It is constructor which takes groupMemberList as parameter and bind it with groupMemberList in adapter.
     *
     * @param context is a object of Context.
     * @param groupMemberList is a list of group member used in this adapter.
     */
    constructor(context: Context, groupMemberList: List<GroupMember>?, groupOwnerId: String?):super(){
        this.groupMemberList = groupMemberList!! as MutableList<GroupMember>
        this.groupOwnerId = groupOwnerId
        this.context = context
        fontUtils = FontUtils.getInstance(context)
    }

    constructor(context: Context) : super() {
        this.context = context
        fontUtils = FontUtils.getInstance(context)
    }


    init {
        updateGroupMembers(groupMemberList!!)
//        this.groupOwnerId = groupOwnerId
//        this.context = context
//        fontUtils = FontUtils.getInstance(context)
    }

    override fun onCreateViewHolder(parent: ViewGroup, i: Int): GroupMemberViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val userListRowBinding: CometchatUserListItemBinding = DataBindingUtil.inflate(layoutInflater, R.layout.cometchat_user_list_item, parent, false)
        return GroupMemberViewHolder(userListRowBinding)
    }

    /**
     * This method is used to bind the GroupMemberViewHolder contents with groupMember at given
     * position. It set name icon, scope with respective GroupMemberViewHolder content.
     *
     * @param groupMemberViewHolder is a object of GroupMemberViewHolder.
     * @param i is a position of item in recyclerView.
     * @see GroupMember
     */
    override fun onBindViewHolder(groupMemberViewHolder: GroupMemberViewHolder, i: Int) {
        val groupMember = groupMemberList[i]
        groupMemberViewHolder.userListRowBinding.avUser.setBackgroundColor(context.resources.getColor(R.color.colorPrimary))
        groupMemberViewHolder.userListRowBinding.root.setTag(R.string.user, groupMember)
        if (groupMember.uid == CometChat.getLoggedInUser().uid) {
            groupMemberViewHolder.userListRowBinding.txtUserName.setText(R.string.you)
        } else groupMemberViewHolder.userListRowBinding.txtUserName.text = groupMember.name
        if (groupOwnerId != null && groupMember.uid == groupOwnerId && groupMember.scope == CometChatConstants.SCOPE_ADMIN) {
            groupMemberViewHolder.userListRowBinding.txtUserScope.setText(R.string.owner)
        } else if (groupMember.scope == CometChatConstants.SCOPE_ADMIN) {
            groupMemberViewHolder.userListRowBinding.txtUserScope.setText(R.string.admin)
        } else if (groupMember.scope == CometChatConstants.SCOPE_MODERATOR) {
            groupMemberViewHolder.userListRowBinding.txtUserScope.setText(R.string.moderator)
        } else {
            groupMemberViewHolder.userListRowBinding.txtUserScope.text = ""
        }
        groupMemberViewHolder.userListRowBinding.txtUserName.typeface = fontUtils.getTypeFace(FontUtils.robotoMedium)
        if (groupMember.avatar == null || groupMember.avatar.isEmpty()) groupMemberViewHolder.userListRowBinding.avUser.setInitials(groupMember.name) else groupMemberViewHolder.userListRowBinding.avUser.setAvatar(groupMember.avatar)
        if (groupMember.status.equals(CometChatConstants.USER_STATUS_ONLINE, ignoreCase = true)) groupMemberViewHolder.userListRowBinding.statusIndicator.visibility = View.VISIBLE

        groupMemberViewHolder.userListRowBinding.statusIndicator.setUserStatus(groupMember.status)

        if (Utils.isDarkMode(context)) {
            groupMemberViewHolder.userListRowBinding.txtUserName.setTextColor(context.resources.getColor(R.color.textColorWhite))
            groupMemberViewHolder.userListRowBinding.tvSeprator.setBackgroundColor(context.resources.getColor(R.color.grey))
            groupMemberViewHolder.userListRowBinding.txtUserScope.setTextColor(context.resources.getColor(R.color.textColorWhite))
        } else {
            groupMemberViewHolder.userListRowBinding.txtUserName.setTextColor(context.resources.getColor(R.color.primaryTextColor))
            groupMemberViewHolder.userListRowBinding.tvSeprator.setBackgroundColor(context.resources.getColor(R.color.light_grey))
            groupMemberViewHolder.userListRowBinding.txtUserScope.setTextColor(context.resources.getColor(R.color.secondaryTextColor))
        }
        if (i == itemCount - 1) groupMemberViewHolder.userListRowBinding.tvSeprator.visibility = View.GONE
    }

    override fun getItemCount(): Int {
        return groupMemberList.size
    }

    /**
     * This method is used to add group members in groupMemberList of adapter
     *
     * @param groupMembers is a list of group members which will be added in adapter.
     * @see GroupMember
     */
    fun addAll(groupMembers: List<GroupMember?>?) {
        for (groupMember in groupMembers!!) {
            if (!groupMemberList.contains(groupMember)) {
                groupMemberList.add(groupMember!!)
            }
        }
        notifyDataSetChanged()
    }

    /**
     * This method is used to set search list in a groupMemberList of adapter.
     *
     * @param filterlist is a list of searched group members.
     */
    fun searchGroupMembers(filterlist: MutableList<GroupMember>) {
        groupMemberList = filterlist
        notifyDataSetChanged()
    }

    /**
     * This method is used to add group member in a groupMemberList of adapter.
     *
     * @param joinedUser is object of GroupMember which will be added in a groupList.
     * @see GroupMember
     */
    fun addGroupMember(joinedUser: GroupMember) {
        groupMemberList.add(joinedUser)
        notifyDataSetChanged()
    }

    /**
     * This method is used to remove group member from groupMemberList of adapter.
     *
     * @param groupMember is a object of GroupMember which will be removed from groupList.
     * @see GroupMember
     */
    fun removeGroupMember(groupMember: GroupMember?) {
        if (groupMemberList.contains(groupMember)) {
            groupMemberList.remove(groupMember)
            notifyDataSetChanged()
        }
    }

    /**
     * This method is used to update group member from a groupMemberList of a adapter.
     *
     * @param groupMember is a object of GroupMember which will updated with old group member in
     * groupMemberList.
     * @see GroupMember
     */
    fun updateMember(groupMember: GroupMember) {
        if (groupMemberList.contains(groupMember)) {
            val index = groupMemberList.indexOf(groupMember)
            groupMemberList.remove(groupMember)
            groupMemberList.add(index, groupMember)
            notifyItemChanged(index)
        }
    }

    fun resetAdapter() {
        groupMemberList.clear()
        notifyDataSetChanged()
    }

    inner class GroupMemberViewHolder(var userListRowBinding: CometchatUserListItemBinding) : RecyclerView.ViewHolder(userListRowBinding.root)

    /**
     * This method is used to update group members in a groupMemberList of a adapter.
     *
     * @param groupMembers is a list of updated group members.
     */
    fun updateGroupMembers(groupMembers: List<GroupMember>) {
        for (i in groupMembers.indices) {
            if (groupMemberList.contains(groupMembers[i])) {
                val index = groupMemberList.indexOf(groupMembers[i])
                groupMemberList.removeAt(index)
                groupMemberList.add(index, groupMembers[i])
            } else {
                groupMemberList.add(groupMembers[i])
            }
        }
        notifyDataSetChanged()
    }

    companion object {
        private val TAG = GroupMemberAdapter::class.java.simpleName
    }
}